File: src/main/java/core/procedures/math/Log.java
package core.procedures.math;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.BigComplex;
import core.scm.BigRational;
import core.utils.Utils;

import java.math.BigDecimal;
import java.math.BigInteger;

public final class Log extends AFn {

  /* If number has 307 digits or less, then can use Math.log(double) */
  private static final int MAX_DIGITS = 307;
  private static final double VALUE = Math.log(Math.pow(10, MAX_DIGITS));

  public Log() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{Number.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "log";
  }

  @Override
  public Number apply1(Object arg) {
    return log((Number)arg);
  }

  public static Number log(Number number) {
    if (number instanceof BigComplex) {
      return ((BigComplex)number).log();
    }
    if (number instanceof Double) {
      return Math.log(number.doubleValue());
    }
    Number n = Utils.upcast(number);
    if (n instanceof Long) {
      if (n.longValue() == 0) {
        throw new ArithmeticException("log: undefined for 0");
      }
      if (n.longValue() == 1) {
        return 0L;
      }
      return Math.log(n.doubleValue());
    }
    if (number instanceof BigRational) {
      if (((BigRational) number).isZero()){
        throw new ArithmeticException("log: undefined for 0");
      }
      if (number.equals(BigRational.ONE)) {
        return 0L;
      }
      return logBig(((BigRational)number).toBigDecimal());
    }
    if (number instanceof BigDecimal) {
      if (((BigDecimal)number).signum() == 0) {
        throw new ArithmeticException("log: undefined for 0");
      }
      return logBig((BigDecimal) number);
    }
    if (number instanceof BigInteger) {
      if (((BigInteger)number).signum() == 0) {
        throw new ArithmeticException("log: undefined for 0");
      }
      return logBig(Utils.toBigDecimal(number));
    }
    return Math.log(number.doubleValue());
  }

  /* Natural logarithm for Big numbers (greater than Double.MAX_VALUE) */
  private static Number logBig(BigDecimal number) {
    if (Double.isFinite(number.doubleValue())) {
      return Math.log(number.doubleValue());
    }
    int digits = integerDigits(number);
    int n = digits / MAX_DIGITS;
    number = number.movePointLeft(n * MAX_DIGITS);
    return (n * VALUE) + Math.log(number.doubleValue());
  }

  /* Return number of digits of a given BigDecimal number */
  private static int integerDigits(BigDecimal n) {
    return n.signum() == 0 ? 1 : n.precision() - n.scale();
  }
}

File: src/main/java/core/procedures/math/Log.kt
package core.procedures.math

import core.procedures.AFn
import core.scm.BigComplex
import core.scm.BigRatio
import core.utils.Utils
import java.math.BigDecimal
import java.math.BigInteger
import kotlin.math.ln
import kotlin.math.pow

class Log : AFn<Number?, Number>(name = "log", isPure = true, minArgs = 1, maxArgs = 1,
                                 mandatoryArgsTypes = arrayOf<Class<*>>(Number::class.java)) {

    companion object {
        /* If number has 307 digits or less, then can use ln(double) */
        private val MAX_DIGITS = 307
        private val VALUE = ln(10.0.pow(MAX_DIGITS.toDouble()))
    }

    override operator fun invoke(arg: Number?) = when {
        Utils.isZero(arg) && Utils.isExact(arg) -> throw ArithmeticException("log: undefined for 0")
        Utils.isOne(arg)  && Utils.isExact(arg) -> 0L
        arg is BigComplex                       -> arg.log()
        arg is BigRatio                         -> logBig(arg.toBigDecimal())
        arg is BigDecimal                       -> logBig(arg)
        arg is BigInteger                       -> logBig(BigDecimal(arg))
        else                                    -> ln(arg!!.toDouble())
    }

    /* Natural logarithm for Big numbers (greater than Double.MAX_VALUE) */
    private fun logBig(number: BigDecimal) = when {
        number.toDouble().isFinite() -> ln(number.toDouble())
        else -> {
            val digits = integerDigits(number)
            val num = number.movePointLeft(digits - digits.rem(MAX_DIGITS))
            (digits / MAX_DIGITS) * VALUE + ln(num.toDouble())
        }
    }

    /* Return number of digits of a given BigDecimal number */
    private fun integerDigits(n: BigDecimal) = if (n.signum() == 0) 1 else n.precision() - n.scale()
}
